### PR TITLE
Add support for MegaBlocks MoEs

### PR DESCRIPTION
**Pull request by tgale96**
These changes add support for using MegaBlocks dMoE and MoE layers in Megatron. MegaBlocks is exposed through an [adapter](https://github.com/NVIDIA/Megatron-LM/compare/main...stanford-futuredata:Megatron-LM:main#diff-aa9d60b130b2ce6bd6810f247a0e1770fe0d0279d01cf8b491cd03df2c72be7a) which isolates the megablocks package dependency so that it does not need to be installed if users are not training MoEs. The changes can be grouped by features:

Basic dMoE/MoE Support:
- Add wrappers for MegaBlocks layers in [megatron/model/transformer.py](https://github.com/NVIDIA/Megatron-LM/compare/main...stanford-futuredata:Megatron-LM:main#diff-46c4c76deb18adf1de8e0be6d4229baed5f1f0308e141479f5a993b3d83dd445)
- Add load balancing loss support in [pretrain_gpt.py](https://github.com/NVIDIA/Megatron-LM/compare/main...stanford-futuredata:Megatron-LM:main#diff-8f7acbd2608d54e2faf8653c0d144c718cd78bcb6a53430c35e81199c6c6651a)
- Add MoE arguments in [megatron/arguments.py](https://github.com/NVIDIA/Megatron-LM/compare/main...stanford-futuredata:Megatron-LM:main#diff-5f7d1ddfb0666cb6bb4ec0f07fd2fd7b1cd0354f421df5560489091db2ff5a55)
- Document MoE support in [README.md](https://github.com/NVIDIA/Megatron-LM/compare/main...stanford-futuredata:Megatron-LM:main#diff-b335630551682c19a781afebcf4d07bf978fb1f8ac04c6bf87428ed5106870f5)

Expert Model Parallelism Support:
- Exclude expert parameters from data parallel gradient reduction in [megatron/model/distributed.py](https://github.com/NVIDIA/Megatron-LM/compare/main...stanford-futuredata:Megatron-LM:main#diff-2f144073023d6fc91a0e4c2264da44f6fcabc365f9e5ff6de7eedc49984ce933)
- Handle expert parameters in [megatron/optimizer/optimizer.py](https://github.com/NVIDIA/Megatron-LM/compare/main...stanford-futuredata:Megatron-LM:main#diff-dabd5ffec3ce00acedf6bf253e236186da42c61e2501d7d3f5a84645bfb6578b)
- Handle sharded expert parameter gradient clipping in [megatron/optimizer/clip_grads.py](https://github.com/NVIDIA/Megatron-LM/compare/main...stanford-futuredata:Megatron-LM:main#diff-70e47541d758dfcccac5b850fe8b902aae8359d14094b1b8ae3b161e5fae2063)
- Handle sharded checkpointing in [megatron/checkpointing.py](https://github.com/NVIDIA/Megatron-LM/compare/main...stanford-futuredata:Megatron-LM:main#diff-122925dfa160fba3c00803abba3577ef0d5aa5ab48989032a63d41c91f2a8002)

Pipeline Parallelism Support:
- Add new [ModelType](https://github.com/NVIDIA/Megatron-LM/compare/main...stanford-futuredata:Megatron-LM:main#diff-ed97fa9d231bf55bfc09a7413d8f54f3df7827cc08f0a446746e60a9a7f9e933) to identify when per-layer load balancing losses are present.
- Handle load balancing loss for pipeline parallelism in [megatron/schedules.py](https://github.com/NVIDIA/Megatron-LM/compare/main...stanford-futuredata:Megatron-LM:main#diff-0b22c62a98249377975dafb92243b230ee5d1ab8c7aef7275d0a2465d669e0cb).
- Handle load balancing loss for pipeline parallelism in [megatron/training.py](https://github.com/NVIDIA/Megatron-LM/compare/main...stanford-futuredata:Megatron-LM:main#diff-e2b248f8c422a601bcb0b7d93f96c1dff070f2694737e2b69f1def64ab9c1844)

This PR adds MegaBlocks MoE support to GPT models. Extending support to other pretraining scripts requires limited changes. Expert model parallelism in this PR is supported across the data parallel group. We're interested in de-coupling this, and also extending MegaBlocks to support tensor model parallelism in future PRs.

### CODE REVIEW

**Comment by tgale96** [comment]
Closing in favor of https://github.com/NVIDIA/Megatron-LM/pull/288, which starts by integrating the changes listed under "Basic dMoE/MoE Support".
